File: 11-Tree/programs/ruby/program3.rb
class Node
    attr_accessor :lchild, :data, :rchild

    def initialize(data)
        @data = data
        @lchild = nil
        @rchild = nil
    end
end

class Queue
    def initialize(size = 10)
        @size = size
        @front = @rear = -1
        @q = Array.new(@size)
    end

    def enqueue(x)
        if @rear == @size - 1
            puts "Queue is full"
        else
            @rear += 1
            @q[@rear] = x
        end
    end

    def dequeue
        if @front == @rear
            puts "Queue is empty"
            nil
        else
            @front += 1
            @q[@front]
        end
    end

    def is_empty
        @front == @rear
    end
end

class Tree
    attr_accessor :root

    def initialize
        @root = nil
    end

    def create_tree
        q = Queue.new(100)
        puts "Enter root value: "
        x = gets.chomp.to_i
        @root = Node.new(x)
        q.enqueue(@root)
        while !q.is_empty
            p = q.dequeue
            puts "Enter left child of #{p.data}: "
            x = gets.chomp.to_i
            if x != -1
                t = Node.new(x)
                p.lchild = t
                q.enqueue(t)
            end
            puts "Enter right child of #{p.data}: "
            x = gets.chomp.to_i
            if x != -1
                t = Node.new(x)
                p.rchild = t
                q.enqueue(t)
            end
        end
    end

    def pre_order(p)
        if p
            print "#{p.data} "
            pre_order(p.lchild)
            pre_order(p.rchild)
        end
    end

    def in_order(p)
        if p
            in_order(p.lchild)
            print "#{p.data} "
            in_order(p.rchild)
        end
    end

    def post_order(p)
        if p
            post_order(p.lchild)
            post_order(p.rchild)
            print "#{p.data} "
        end
    end

    def level_order(p)
        q = Queue.new(100)
        print "#{@root.data} "
        q.enqueue(@root)
        while !q.is_empty
            p = q.dequeue
            if p.lchild
                print "#{p.lchild.data} "
                q.enqueue(p.lchild)
            end
            if p.rchild
                print "#{p.rchild.data} "
                q.enqueue(p.rchild)
            end
        end
    end

    def height(p)
        return 0 if p.nil?

        x = height(p.lchild)
        y = height(p.rchild)
        x > y ? x + 1 : y + 1
    end
end

t = Tree.new
t.create_tree
t.pre_order(t.root)
puts
t.in_order(t.root)
puts
t.post_order(t.root)
puts
t.level_order(t.root)
puts
puts t.height(t.root)

File: 11-Tree/programs/cpp/program3.cpp
// Creating binary tree in cpp with using class


#include<bits/stdc++.h>
using namespace std;

class Node{
  public:
    Node *lchild;
    int data;
    Node *rchild;
};

class Queue{
  private:
    int size;
    int front;
    int rear;
    Node **Q;
  public:
    Queue(){
      size = 10;
      front = rear = -1;
      Q = new Node*[this->size];
    }
    Queue(int size){
      this->size = size;
      front = rear = -1;
      Q = new Node*[this->size];
    }
    void enqueue(Node* x);
    Node * dequeue();
    void Display();
    int isEmpty();
};

void Queue::enqueue(Node* x){
  if(rear == size-1){
    cout << "Queue is full" << endl;
  }else{
    rear++;
    Q[rear] = x;
  }
}

Node* Queue::dequeue(){
  Node* x = NULL;
  if(front == rear){
    cout << "Queue is empty" << endl;
  }else{
    front++;
    x = Q[front];
  }
  return x;
}

void Queue::Display(){
  for(int i = front+1; i <= rear; i++){
    cout << Q[i] << " ";
  }
  cout << endl;
}

int Queue::isEmpty(){
  return front == rear;
}

class Tree{
  public:
    Node *root;
    Tree(){
      root = NULL;
    }
    void createTree();
    void preOrder(Node *p);
    void inOrder(Node *p);
    void postOrder(Node *p);
    void levelOrder(Node *p);
    int height(Node *p);
};

void Tree::createTree(){
  Node *p, *t;
  int x;
  Queue q(100);
  cout << "Enter root value: ";
  cin >> x;
  cout << x << endl;
  root = new Node;
  root->data = x;
  root->lchild = root->rchild = NULL;
  q.enqueue(root);
  while(!q.isEmpty()){
    p = q.dequeue();
    cout << "Enter left child of " << p->data << ": ";
    cin >> x;
    cout << x << endl;
    if(x != -1){
      t = new Node;
      t->data = x;
      t->lchild = t->rchild = NULL;
      p->lchild = t;
      q.enqueue(t);
    }
    cout << "Enter right child of " << p->data << ": ";
    cin >> x;
    cout << x << endl;
    if(x != -1){
      t = new Node;
      t->data = x;
      t->lchild = t->rchild = NULL;
      p->rchild = t;
      q.enqueue(t);
    }
  }
} 

void Tree::preOrder(Node *p){
  if(p){
    cout << p->data << " ";
    preOrder(p->lchild);
    preOrder(p->rchild);
  }
}

void Tree::inOrder(Node *p){
  if(p){
    inOrder(p->lchild);
    cout << p->data << " ";
    inOrder(p->rchild);
  }
}

void Tree::postOrder(Node *p){
  if(p){
    postOrder(p->lchild);
    postOrder(p->rchild);
    cout << p->data << " ";
  }
}

void Tree::levelOrder(Node *p){
  Queue q(100);
  cout << root->data << " ";
  q.enqueue(root);
  while(!q.isEmpty()){
    p = q.dequeue();
    if(p->lchild){
      cout << p->lchild->data << " ";
      q.enqueue(p->lchild);
    }
    if(p->rchild){
      cout << p->rchild->data << " ";
      q.enqueue(p->rchild);
    }
  }
}

int Tree::height(Node *p){
  int x = 0, y = 0;
  if(p == NULL){
    return 0;
  }
  x = height(p->lchild);
  y = height(p->rchild);
  if(x > y){
    return x+1;
  }else{
    return y+1;
  }
}

int main(){
  Tree t;
  t.createTree();
  t.preOrder(t.root);
  cout << endl;
  t.inOrder(t.root);
  cout << endl;
  t.postOrder(t.root);
  cout << endl;
  t.levelOrder(t.root);
  cout << endl;
  cout << t.height(t.root) << endl;
  
  return 0;
}
// Output:
// Enter root value: 1
// 1
// Enter left child of 1: 2
// 2
// Enter right child of 1: 3
// 3
// Enter left child of 2: 4
// 4
// Enter right child of 2: 5
// 5
// Enter left child of 3: 6
// 6
// Enter right child of 3: 7
// 7
// 1 2 4 5 3 6 7
